packages/@aws-cdk/cdk/lib/aspects/tag-aspect.ts

```diff
@@ -0,0 +1,149 @@
+    if (this.visitedBy[node.uniqueId] === true) {
+      return;
+    }
+    this.visitedBy[node.uniqueId] = true;
+    const propagate = this.propagate > 0 ? this.propagate - 1 : this.propagate;
+    this.parent = this.parent || node;
+    if (propagate !== 0) {
+      for (const child of node.children) {
+        this.propagate = propagate;
+        child.apply(this);
+      }
+    }
+    if (node instanceof Resource && (node as Resource).tagType !== TagType.NotTaggable) {
+      const resource = node as Resource;
+      if (this.exclude.length !== 0 && this.exclude.indexOf(resource.resourceType) !== -1) {
+        return;
+      }
+      if (this.include.length !== 0 && this.include.indexOf(resource.resourceType) === -1) {
+        return;
+      }
+      if (this.hasPrecedence(node.aspects)) {
+        this.applyTag(resource);
+      }
+    }
+  }
+
+  protected abstract applyTag(resource: Resource): void;
+
+  private hasPrecedence(aspects: IAspect[]): boolean {
+    const tagAspects = aspects.filter( a => a !== this && a.type === this.type);
+    const collisions = tagAspects.filter( a => (a as TagBaseAspect).key === this.key);
+    for (const collidingAspect of collisions) {
+      const collision = collidingAspect as TagBaseAspect;
+      // if the parent is not defined the apsect hasn't been visited and is
+      // local to the current node; thus the collision has precedence because it
+      // is later (e.g. last one in wins) in the array
+      if (collision.parent === undefined) {
+        return false;
+      }
+      // the aspect with the most ancestors is closest to this node in the tree
+      if (collision.parent.ancestors().length > this.parent!.ancestors().length) {
+        return false;
+      }
+    }
+    return true;
+  }
+}
+
+export class TagAspect extends TagBaseAspect {
+
+  constructor(key: string, value: string, props: TagProperties = {}) {
+    super(key, {value, ...props});
+    if (this.value === undefined) {
+      throw new Error('Tag must have a value');
+    }
+  }
+
+  protected applyTag(resource: Resource) {
+    removeTag(this, resource);
+    setTag(this, resource);
+  }
+}
+
+export class RemoveTag extends TagBaseAspect {
+
+  constructor(key: string, props: TagProperties = {}) {
+    super(key, props);
+  }
+
+  protected applyTag(resource: Resource) {
+    return removeTag(this, resource);
+  }
+}
+
+function removeTag(tag: TagBaseAspect, resource: Resource): void {
+  const key = tag.key;
+  switch (resource.tagType) {
+    case TagType.Standard:
+    case TagType.AutoScalingGroup: {
+      if (resource.properties.tags === undefined) { break; }
+      resource.properties.tags = resource.properties.tags.filter( (t: any ) => t.key !== key);
+      break;
+    }
+    case TagType.Map: {
+      if (resource.properties.tags === undefined) { break; }
+      delete resource.properties.tags[key];
+      break;
+    }
+  }
+}
+
+function setTag(tag: TagBaseAspect, resource: Resource) {
```

Comment:
Exactly this function is why I wanted to keep `TagManager`. Now, TagManager doesn't need to do complicated work anymore, but the type-specific information about how a tag should be rendered can be kept there.

On the other hand, if we're sure `TagType` does not need to be extensible... I guess okay.

I would prefer for `setTag()` to either take the actual tag key and value if you're going to keep it a free function, or become a member function of `TagBaseAspect` and use `this`.

Also, I would prefer for the second call to `setTag()` to overwrite the first, if present.

Reply:
I think this moves to base, but let me see if removing propagate changes my opinion.

> Also, I would prefer for the second call to setTag() to overwrite the first, if present.
You mean as opposed to my use of delete then create? Find the index and just replace if there is a second call, is what you would prefer?

Reply:
Yes.